### DISCUSSION TITLE
My realtime subscriptions get terminated and I cannot recover them

### DISCUSSION THREAD

**Discussion by ivanbrykov** [supabase/supabase#5312] [Questions]
I am working on an app that is supposed to stay in the browser tab forever. User is not supposed to re-open or reload it all the time. At the moment I maintain 2 realtime subscriptions. My problem is that after some time of inactivity (laptop's lid closed) when I get back to the app I see that my subscriptions are gone and I cannot re-establish them unless I reload the page. I added a bit of logging to the app and here is what it looks like:
![image](https://user-images.githubusercontent.com/476516/152475628-c005fd8f-d97b-43d4-8300-1fcea51e1370.png)

As you can see, I start by invoking the ```supabase.getSubscriptions()``` method which returns an empty array (no subscriptions).
I invoke ```supabase.auth.refreshSession()``` hoping that it will allow me to re-subscribe.
And it does — I have 2 subscriptions right after that.
But after about 8 seconds they get closed and I end up in the square 0.

This is cloud supabase not self-hosted. 
Browser: MacOS Safari

I would appreciate any comments or solutions to this problem, thank you!

**Comment by GaryAustin1** [comment]
There was an issue with token change and realtime.js a few weeks ago but that was fixed.  Make sure you have latest drivers.
You should look at the actual websocket responses to see the error and why it closed.

BUT, What you want to do with realtime won't work (at least currently).  If the internet connection drops OR the browser tab goes into the background (including power down mode) after 5 minutes safari and edge (at least on Windows) throttle the timers and cause realtime heartbeat to drop.  This can occur with no error as it retries the connection for awhile.  I have code to detect loss of connection or visibility of the browser tab and shut down realtime gracefully after 5 minutes and then reload my needed table info before restarting it when connection/tab comes back.  Realtime does not queue so even on a slight internet disconnect you could lose realtime changes.

**Reply by ivanbrykov** [reply]
Thank you for your input!
When you say shut down gracefully, do you mean you invoke removeAllSubscriptions() or something to that effect?
An when you say reload, do you mean force reloading of the page?

**Reply by GaryAustin1** [reply]
Edit:  I clarified what I do on disconnect versus what I said before edit!
Not a page reload, a query to refresh(reload) data.

I remove all subscriptions on page/tab going into background (use the browser visibility handler) for more than 5 minutes.  Then on page coming out of background I restart connections and do a table query to reload any information that might have been missed while connection was gone.  I also reload on disconnect/reconnect of the socket monitoring status and reload table on reconnect.

Note there could also be a bug in supabase.js refreshing the token to realtime when you do the supabase.auth.refreshSession... but it does trigger off of auth event token change, which you get.

I suspect you would see a token error in the websocket responses if this is the case.  On older supabase.js it would fail like you see.

**Reply by emileond** [reply]
I'm experiencing a similar problem @brykov did you find a workaround for this?

**Reply by ivanbrykov** [reply]
> I'm experiencing a similar problem @brykov did you find a workaround for this?

No, I did not find a workaround. My decision was to abandon the real time functionality and just rely on my own refresh cadence.

**Reply by GaryAustin1** [reply]
> I'm experiencing a similar problem @brykov did you find a workaround for this?

@emileond I posted this shortly after this thread.  It seems to work for me for the devices/browser I've tested...

How to obtain reliable realtime updates in the real world
https://github.com/supabase/supabase/discussions/5641

**Reply by emileond** [reply]
Thanks @GaryAustin1 that was very insightful 🙌,

In my case, the subscription status changed to "CLOSED" when the tab was not visible, but the actual connection state remained open, so Supabase never attempted to resubscribe when the tab became visible again.

I ended up manually calling'supabase.realtime.disconnect()' when the subscription status changed to "CLOSED," just to make sure the connection was closing, and it appears to work. I'll probably add visibility event listeners to gain more control - Just sharing this here in case it's useful to someone.

**Reply by GaryAustin1** [reply]
@emileond  that is odd.  I've never seen it close and not resubscribe.  I'll try and run my test again as maybe something has changed in realtime code.  My current app probably would not notice as soon as I get closed I do my thing and shut it down and restart when visible.

Edit.  I may have posted too quickly on that.  On mobile devices, it does all kind of things

**Comment by t0mdicks0n** [comment]
I am having big issues with this by the end of 2024. It's strange to me that Supabase would advertise "real-time" when it only works for about 8 seconds in an active window before it disconnects. I am trying to catch the disconnects and reconnect but after a couple of hours of tinkering I'm not getting it to work reliably. I just gave up and plugged in Firebase Realtime Database instead. Sharing with the hopes of Supabase investing more time into the real-time feature since it would be awesome to get rid of Firebase completely in modern web dev.

**Reply by schvenk** [reply]
My experience has been similar. On an earlier web project, I was sort-of able to figure out how to reopen closed subscriptions, though I don't think it was perfect and there wasn't really any documentation to help.

Now, on a mobile (Flutter) project, I see subscriptions constantly closing and can't figure out how to reopen them without creating infinite loops—and even if I could, I think it would potentially be less efficient than just polling the database every 5s.

Combined with the Flutter SDK's lack of offline caching and the need to manually handle many-to-many writes, I too am probably going to switch to Firestore. Also looking at Brick and Pocketbase, but Firebase seems like it offers the most out of the box—at (unfortunately) the cost of relational data.

**Reply by t0mdicks0n** [reply]
Thanks for sharing. My experience with Supabase has been flawless except for real-time. I also ran into infinite loops all the time when trying to hack it by reopening connections. In the past 8 years I've never had these same issues on the Firestore/Firebase client libraries or with Node servers running Socket.io. My only hope is that Supabase sees this and fix it.

**Reply by SeraphMX** [reply]
Hi... i had the same problem with my app , but i found a solution. You need a SetTimeOut in onAuthStateChange, here is the example: 

```
 const { data: { subscription } } = supabase.auth.onAuthStateChange(async (_, session) => {
  if (session?.user) {
    setTimeout(async () => {
      // Call database here !
      const { data } = await supabase
          .from('users')
          .select('*')
          .eq('id', session.user.id)
          .single();
    })
  }
});
```

For me worked without problems , here is the issue where i found the solution.

[https://github.com/nuxt-modules/supabase/issues/273](https://github.com/nuxt-modules/supabase/issues/273)

**Reply by PanSalut** [reply]
I have the same problem - my Flutter app, after several backgrounds or locking the screen, lost the realtime connection, and I didn't know how to reconnect.

I decided to use Pusher for realtime support, leaving the rest of the backend on Supabase.

I didn't want this change, but it seems to be necessary.

**Comment by PanSalut** [comment]
I had the same problem and was going crazy, but as I can see - it's a Supabase problem.

I'm creating an app that uses chat, I'm using Supabase and Flutter for that.

During testing, I noticed that Supabase closes the realtime connection when the screen locks or the app goes to the background. Then, it's hard to restore the connection.

I decided to use Pusher and so far - everything works great. The connection is automatically renewed when returning to the app, there are no problems with accidental disconnections from realtime channels. An additional plus is that I can connect to several channels at the same time.

I love Supabase, but I can't stand this Realtime. I hope that this problem will be solved in the future.

**Comment by skelmen** [comment]
I also encounter the same Supabase Realtime disconnection issue on minimize the app to the background or screen locking. I'm developing realtime words game and users connection is very significant part. I am very upset I should refactor my code to use similar services like Pusher or Firebase.

**Comment by Megha30501** [comment]
I am also facing the same issue.

I am building an app that features message chat. The connection works fine for a limited time, but after a few moments, or if the tab is inactive for a while or the laptop goes into sleep mode, I lose the connection and the user has to refresh the page to receive new messages.

**Reply by GaryAustin1** [reply]
Yes this is expected in power saving modes that slow down the clock.  You need to handle errors/visibility or look at the new webworker option to keep it alive...
Around this link about the issue is a solution for some cases:  https://github.com/supabase/realtime-js/issues/121#issuecomment-2629508525

**Reply by Megha30501** [reply]
Thank you, @GaryAustin1, for the help.
